### ISSUE TITLE
populate table using api

### ISSUE THREAD

**Issue by lie16** [bosskmk/pluto_grid#284]
To populate pluto table using json do I need to convert from my custom data model to List<PlutoRow>?

**Comment by anuragowly7** [comment]
can anyone help me with converting json to rows n column objects for plutogrid?

**Comment by anuragowly7** [comment]
> There is currently no automatic conversion feature, so you have to manually convert Json to PlutoRow and PlutoColumn objects.

can anyone help me with converting json to rows n column objects for plutogrid?

**Comment by bosskmk** [comment]
@anuragowly7 
Here is some example.

```dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:pluto_grid/pluto_grid.dart';

class EmptyScreen extends StatefulWidget {
  static const routeName = 'empty';

  const EmptyScreen({Key? key}) : super(key: key);

  @override
  _EmptyScreenState createState() => _EmptyScreenState();
}

class _EmptyScreenState extends State<EmptyScreen> {
  final List<PlutoColumn> columns = [];

  final List<PlutoRow> rows = [];

  late PlutoGridStateManager stateManager;

  @override
  void initState() {
    super.initState();

    const String json =
        '[{"_id": "61c4b1a549f44bb601138297","age": 34,"name": "Kinney Anthony","gender": "male","company": "SCENTY"},{"_id": "61c4b1a53ae228cc32800a51","age": 25,"name": "Barton Hoffman","gender": "female","company": "BIOHAB"}]';

    final parsedJson = jsonDecode(json);

    columns.addAll([
      PlutoColumn(
        title: 'id',
        field: '_id',
        type: PlutoColumnType.text(),
      ),
      PlutoColumn(
        title: 'age',
        field: 'age',
        type: PlutoColumnType.number(),
      ),
      PlutoColumn(
        title: 'name',
        field: 'name',
        type: PlutoColumnType.text(),
      ),
      PlutoColumn(
        title: 'gender',
        field: 'gender',
        type: PlutoColumnType.select(<String>['male', 'female']),
      ),
      PlutoColumn(
        title: 'company',
        field: 'company',
        type: PlutoColumnType.text(),
      ),
    ]);

    for (var item in parsedJson) {
      final Map<String, PlutoCell> cells = {
        for (var e in item.entries) e.key: PlutoCell(value: e.value)
      };

      rows.add(PlutoRow(cells: cells));
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        padding: const EdgeInsets.all(15),
        child: PlutoGrid(
          columns: columns,
          rows: rows,
          onChanged: (PlutoGridOnChangedEvent event) {
            print(event);
          },
          onLoaded: (PlutoGridOnLoadedEvent event) {
            stateManager = event.stateManager;
          },
          configuration: const PlutoGridConfiguration(),
        ),
      ),
    );
  }
}
```

**Comment by anuragowly7** [comment]
Thank You So Much :)

> @anuragowly7 Here is some example.
> 
> ```dart
> import 'dart:convert';
> 
> import 'package:flutter/material.dart';
> import 'package:pluto_grid/pluto_grid.dart';
> 
> class EmptyScreen extends StatefulWidget {
>   static const routeName = 'empty';
> 
>   const EmptyScreen({Key? key}) : super(key: key);
> 
>   @override
>   _EmptyScreenState createState() => _EmptyScreenState();
> }
> 
> class _EmptyScreenState extends State<EmptyScreen> {
>   final List<PlutoColumn> columns = [];
> 
>   final List<PlutoRow> rows = [];
> 
>   late PlutoGridStateManager stateManager;
> 
>   @override
>   void initState() {
>     super.initState();
> 
>     const String json =
>         '[{"_id": "61c4b1a549f44bb601138297","age": 34,"name": "Kinney Anthony","gender": "male","company": "SCENTY"},{"_id": "61c4b1a53ae228cc32800a51","age": 25,"name": "Barton Hoffman","gender": "female","company": "BIOHAB"}]';
> 
>     final parsedJson = jsonDecode(json);
> 
>     columns.addAll([
>       PlutoColumn(
>         title: 'id',
>         field: '_id',
>         type: PlutoColumnType.text(),
>       ),
>       PlutoColumn(
>         title: 'age',
>         field: 'age',
>         type: PlutoColumnType.number(),
>       ),
>       PlutoColumn(
>         title: 'name',
>         field: 'name',
>         type: PlutoColumnType.text(),
>       ),
>       PlutoColumn(
>         title: 'gender',
>         field: 'gender',
>         type: PlutoColumnType.select(<String>['male', 'female']),
>       ),
>       PlutoColumn(
>         title: 'company',
>         field: 'company',
>         type: PlutoColumnType.text(),
>       ),
>     ]);
> 
>     for (var item in parsedJson) {
>       final Map<String, PlutoCell> cells = {
>         for (var e in item.entries) e.key: PlutoCell(value: e.value)
>       };
> 
>       rows.add(PlutoRow(cells: cells));
>     }
>   }
> 
>   @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       body: Container(
>         padding: const EdgeInsets.all(15),
>         child: PlutoGrid(
>           columns: columns,
>           rows: rows,
>           onChanged: (PlutoGridOnChangedEvent event) {
>             print(event);
>           },
>           onLoaded: (PlutoGridOnLoadedEvent event) {
>             stateManager = event.stateManager;
>           },
>           configuration: const PlutoGridConfiguration(),
>         ),
>       ),
>     );
>   }
> }
> ```

ThankYou So Much 💯

**Comment by anuragowly7** [comment]
> There is currently no automatic conversion feature, so you have to manually convert Json to PlutoRow and PlutoColumn objects.

any news on implementing through Json API??

**Comment by anuragowly7** [comment]
//HERE THIS WORKED IF ANYONE NEED IT//

import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import 'package:pluto_grid/pluto_grid.dart';

class Summary {
  final int mID;
  final int eID;
  final String employeeName;
  final String station;
  final String department;
  final String designation;
  final DateTime inTime;
  final String earlyBeforeLate;
  final DateTime outTime;
  final DateTime currentdate;
  final String duration;
  final String processedBy;
  final bool attendeanceStatus;
  final String action;

  Summary({
    required this.mID,
    required this.eID,
    required this.employeeName,
    required this.station,
    required this.department,
    required this.designation,
    required this.inTime,
    required this.earlyBeforeLate,
    required this.outTime,
    required this.currentdate,
    required this.duration,
    required this.processedBy,
    required this.attendeanceStatus,
    required this.action,
  });

  factory Summary.fromJson(Map<String, dynamic> json) {
    return Summary(
      mID: json['MID'],
      eID: json['EID'],
      employeeName: json['Employee_Name'],
      station: json['Station'],
      department: json['Department'],
      designation: json['Designation'],
      inTime: DateFormat.Hms().parse((json['In_Time'])),
      earlyBeforeLate: json['Early_Before_Late'],
      outTime: DateFormat.Hms().parse(json['Out_time']).toLocal(),
      currentdate: DateTime.parse(json['Currentdate']),
      duration: json['Duration'],
      processedBy: json['Processed_By'],
      attendeanceStatus: json['Attendeance_Status'],
      action: json['Action'],
    );
  }
}

Future<List<Summary>> fetchSummary() async {
  final response = await http
      .get(Uri.parse(('https://json.extendsclass.com/bin/d9573a09bf96')));

  if (response.statusCode == 200) {
    var parsed = json.decode(response.body);
    List jsonResponse = parsed["HR"] as List;

    return jsonResponse.map((job) => new Summary.fromJson(job)).toList();
  } else {
    print('Error, Could not load Data.');
    throw Exception('Failed to load Data');
  }
}

class DetailsScreen extends StatefulWidget {
  @override
  _DetailsScreenState createState() => _DetailsScreenState();
}

class _DetailsScreenState extends State<DetailsScreen> {
  final controller = ScrollController();
  double offset = 0;

  late Future<List<Summary>> _func;

  @override
  void initState() {
    _func = fetchSummary();
    controller.addListener(onScroll);
    super.initState();
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  void onScroll() {
    setState(() {
      offset = (controller.hasClients) ? controller.offset : 0;
    });
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: 800,
      width: 1500,
      child: FutureBuilder<List<Summary>>(
        future: _func,
        builder: (ctx, snapshot) {
          if (snapshot.hasData) {
            List<Summary>? data = snapshot.data;
            return PlutoGrid(
              columns: [
                PlutoColumn(
                    title: 'MID',
                    field: 'mid',
                    type: PlutoColumnType.number(),
                    readOnly: true,
                    width: 100,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'EID',
                    field: 'eid',
                    type: PlutoColumnType.number(),
                    readOnly: true,
                    width: 100,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Date',
                    field: 'date',
                    type: PlutoColumnType.date(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Employee Name',
                    field: 'emp_name',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Station',
                    field: 'station',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Department',
                    field: 'department',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Designation',
                    field: 'designation',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'In Time',
                    field: 'in_time',
                    type: PlutoColumnType.time(),
                    readOnly: true,
                    enableColumnDrag: false,
                    width: 100),
                PlutoColumn(
                    title: 'Early Before / Late By',
                    field: 'mean_time',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                  title: 'Out Time',
                  field: 'out_time',
                  type: PlutoColumnType.time(),
                  readOnly: true,
                  enableColumnDrag: false,
                  width: 100,
                ),
                PlutoColumn(
                    title: 'Early Before / Late By',
                    field: 'mean_time2',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Duration',
                    field: 'duration',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Processed By',
                    field: 'processedby',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                    title: 'Status',
                    field: 'status',
                    type: PlutoColumnType.text(),
                    readOnly: true,
                    enableColumnDrag: false),
                PlutoColumn(
                  title: 'Action',
                  field: 'action',
                  type: PlutoColumnType.text(),
                  readOnly: true,
                  enableColumnDrag: false,
                  renderer: (rendererContext) {
                    return Container(
                      child: Row(
                        children: <Widget>[
                          Expanded(
                            child: ElevatedButton(
                              onPressed: () {},
                              child: Text(
                                "Approve",
                              ),
                            ),
                          ),
                          Expanded(
                            child: ElevatedButton(
                              child: Text("Reject"),
                              onPressed: () {},
                              style:
                                  ElevatedButton.styleFrom(primary: Colors.red),
                            ),
                          )
                        ],
                      ),
                    );
                  },
                ),
              ],
              rows: data!
                  .map((country) => PlutoRow(
                        cells: {
                          'mid': PlutoCell(value: country.mID),
                          'eid': PlutoCell(value: country.eID),
                          'emp_name':
                              PlutoCell(value: country.employeeName.toString()),
                          'date': PlutoCell(value: country.currentdate),
                          'station':
                              PlutoCell(value: country.station.toString()),
                          'department':
                              PlutoCell(value: country.department.toString()),
                          'designation':
                              PlutoCell(value: country.designation.toString()),
                          'in_time':
                              PlutoCell(value: country.inTime.toString()),
                          'mean_time': PlutoCell(
                              value: country.earlyBeforeLate.toString()),
                          'out_time':
                              PlutoCell(value: country.outTime.toString()),
                          'mean_time2': PlutoCell(
                              value: country.earlyBeforeLate.toString()),
                          'duration':
                              PlutoCell(value: country.duration.toString()),
                          'processedby':
                              PlutoCell(value: country.processedBy.toString()),
                          'status': PlutoCell(
                              value: country.attendeanceStatus.toString()),
                          'action': PlutoCell(value: country.action.toString()),
                        },
                      ))
                  .toList(),
              onLoaded: (PlutoGridOnLoadedEvent event) {
                event.stateManager.setShowColumnFilter(true);
              },
              onChanged: (PlutoGridOnChangedEvent event) {
                print(event);
              },
              configuration: PlutoGridConfiguration(),
              createFooter: (stateManager) {
                stateManager.setPageSize(10, notify: false); // default 40
                return PlutoPagination(stateManager);
              },
            );
          } else if (snapshot.hasError) {
            return AlertDialog(
              title: Text(
                'An Error Occured!',
                textAlign: TextAlign.center,
                style: TextStyle(
                  color: Colors.redAccent,
                ),
              ),
              content: Text(
                "${snapshot.error}",
                style: TextStyle(
                  color: Colors.blueAccent,
                ),
              ),
              actions: <Widget>[
                ElevatedButton(
                  child: Text(
                    'Go Back',
                    style: TextStyle(
                      color: Colors.redAccent,
                    ),
                  ),
                  onPressed: () {
                    Navigator.of(context).pop();
                  },
                )
              ],
            );
          }
          // By default, show a loading spinner.
          return Center(
            child: Column(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                CircularProgressIndicator(),
                SizedBox(height: 20),
                Text('This may take some time..')
              ],
            ),
          );
        },
      ),
    );
  }
}

**Comment by henry2man** [comment]
@bosskmk When dealing with this Feature Request, please consider some issues related with pagination.

Normally is far from optimal fetch all registries; instead you'll should always paginate results. For example, when you have a Spring Boot/Spring REST API backend service you'll have a built-in pagination mechanism that let's you specify variables like `pageSize`, `pageNumber` and also `sorting` ( https://docs.spring.io/spring-data/rest/docs/current/reference/html/#paging-and-sorting ). 

Nowadays, in order to develop a performant pagination with this kind of APIs you'll need to:

- Disable Pluto_grid built-in pagination mechanism (or simply not use it).
- Disable Pluto_Grid sorting (because there aren't, AFAIK, any callbacks you can listen to in order to repeat backend API calls with newer sort configuration)
- Develop some custom UI&code for sorting and pagination (which is a shame because `pluto_grid` already integrates these visual components)

# What I'm proposing?

- Add a new Dropdown selector, allowing changing `pageSize` dynamically
- Develop new callbacks for paging & sorting integrated widgets, so default sorting & paging behaviour could be overridden (in this case, by making new backend calls)

**Comment by henry2man** [comment]
Well, after some researching I've found that `stateManager` listener could make the trick. For now this is a little bit overwhelming (you'll have only one listener for all grid related events) so every listener will need to compare current `stateManager` values with every corresponding state management variables.

So, in order to:

- Populate table using an API -> Write custom code that translates API responses into `PlutoColumn`'s & `PlutoRow`'s
- Implement server side pagination -> Make your API call && also a second API call in order to get total number of results. Then create a "results" list, adding the first page with real data and mocking the rest with empty values. Using `stateManager` listeners you'll need to lazily fetch data when page is changed ( NOTE: It is key here that the listeners are called BEFORE the page change or else the pagination might fail. )
- Implement server side sorting & filtering -> The same as pagination, a `stateManager` listener must be setup but the listener must be called BEFORE actually sorting & filtering occurs.

**Comment by MalikSamiAwan** [comment]
> Well, after some researching I've found that `stateManager` listener could make the trick. For now this is a little bit overwhelming (you'll have only one listener for all grid related events) so every listener will need to compare current `stateManager` values with every corresponding state management variables.
> 
> So, in order to:
> 
> * Populate table using an API -> Write custom code that translates API responses into `PlutoColumn`'s & `PlutoRow`'s
> * Implement server side pagination -> Make your API call && also a second API call in order to get total number of results. Then create a "results" list, adding the first page with real data and mocking the rest with empty values. Using `stateManager` listeners you'll need to lazily fetch data when page is changed ( NOTE: It is key here that the listeners are called BEFORE the page change or else the pagination might fail. )
> * Implement server side sorting & filtering -> The same as pagination, a `stateManager` listener must be setup but the listener must be called BEFORE actually sorting & filtering occurs.

could you please share an example

**Comment by bosskmk** [comment]
Released PlutoGrid 5.3.0

Refer to below link.
https://github.com/bosskmk/pluto_grid/issues/601
